tests/pubsub/testgossipmembership.nim

```diff
@@ -0,0 +1,240 @@
+
+  # Generalized setup function to initialize one or more topics
+  proc setupGossipSub(
+      topics: seq[string], numPeers: int
+  ): (TestGossipSub, seq[Connection]) =
+    let gossipSub = TestGossipSub.init(newStandardSwitch())
+    var conns = newSeq[Connection]()
+
+    for topic in topics:
+      gossipSub.mesh[topic] = initHashSet[PubSubPeer]()
+      gossipSub.topicParams[topic] = TopicParams.init()
+      gossipSub.gossipsub[topic] = initHashSet[PubSubPeer]()
+
+      for i in 0 ..< numPeers:
+        let conn = TestBufferStream.new(noop)
+        conns &= conn
+        let peerId = randomPeerId()
+        conn.peerId = peerId
+        let peer = gossipSub.getPubSubPeer(peerId)
+        peer.sendConn = conn
+        gossipSub.gossipsub[topic].incl(peer)
+
+    return (gossipSub, conns)
+
+  # Wrapper function to initialize a single topic by converting it into a seq
+  proc setupGossipSub(topic: string, numPeers: int): (TestGossipSub, seq[Connection]) =
+    setupGossipSub(@[topic], numPeers)
+
+  # Helper function to subscribe to topics
+  proc subscribeToTopics(gossipSub: TestGossipSub, topics: seq[string]) =
+    for topic in topics:
+      gossipSub.subscribe(
+        topic,
+        proc(topic: string, data: seq[byte]): Future[void] {.async.} =
+          discard
+        ,
+      )
+
+  # Helper function to unsubscribe to topics
+  proc unsubscribeFromTopics(gossipSub: TestGossipSub, topics: seq[string]) =
+    for topic in topics:
+      gossipSub.unsubscribeAll(topic)
+
+  proc commonSubscribe(
+      nodes: seq[TestGossipSub],
+      topic: string,
+      handler: proc(topic: string, data: seq[byte]) {.async.},
+  ) =
+    for node in nodes:
+      node.subscribe(topic, handler)
+    echo "Subscribed all nodes to the topic: ", topic
+
+  proc commonUnsubscribe(
+      nodes: seq[TestGossipSub],
+      topic: string,
+      handler: proc(topic: string, data: seq[byte]) {.async.},
+  ) =
+    for node in nodes:
+      node.unsubscribe(topic, handler)
+    echo "Unsubscribed all nodes from the topic: ", topic
+
+  # Simulate the `SUBSCRIBE` to the topic and check proper handling in the mesh and gossipsub structures
+  asyncTest "handle SUBSCRIBE to the topic":
+    let topic = "test-topic"
+    let (gossipSub, conns) = setupGossipSub(topic, 5)
+
+    subscribeToTopics(gossipSub, @[topic])
+
+    check gossipSub.topics.contains(topic)
+
+    check gossipSub.gossipsub[topic].len() == 5
+
+    await allFuturesThrowing(conns.mapIt(it.close()))
+    await gossipSub.switch.stop()
+
+  # Simulate an UNSUBSCRIBE to the topic and check if the topic is removed from the relevant data structures but remains in gossipsub
+  asyncTest "handle UNSUBSCRIBE to the topic":
+    let topic = "test-topic"
+    let (gossipSub, conns) = setupGossipSub(topic, 5)
+
+    subscribeToTopics(gossipSub, @[topic])
+
+    unsubscribeFromTopics(gossipSub, @[topic])
+
+    check topic notin gossipSub.topics
```

Comment:
For the cases where the value exists (`topic in gossipSub.gossipsub`), do make the check in a more exhaustive way. That is, checking for length as you did before.